File: sudoku-web/lib/remarkable.ts
import JSZip from "jszip";

const AUTH = "https://my.remarkable.com";
const STORAGE = "https://document-storage-production-dot-remarkable-production.appspot.com/document-storage/json/2";

function nowIso(): string {
  return new Date().toISOString().replace(/(\.\d{3})\d*Z/, "$1Z");
}

async function check(res: Response, label: string): Promise<void> {
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`${label} failed (${res.status}): ${body.slice(0, 200)}`);
  }
}

export async function registerDevice(code: string): Promise<string> {
  const res = await fetch(`${AUTH}/token/json/2/device/new`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      code: code.trim().toLowerCase(),
      deviceDesc: "desktop-linux",
      deviceID: crypto.randomUUID(),
    }),
  });
  await check(res, "Device registration");
  return res.text();
}

async function sessionToken(deviceToken: string): Promise<string> {
  const res = await fetch(`${AUTH}/token/json/2/user/new`, {
    method: "POST",
    headers: { Authorization: `Bearer ${deviceToken}` },
  });
  await check(res, "Session token");
  return res.text();
}

export async function uploadPdf(
  pdfBytes: Uint8Array,
  name: string,
  deviceToken: string
): Promise<string> {
  const session = await sessionToken(deviceToken);
  const authHeaders = { Authorization: `Bearer ${session}` };
  const docId = crypto.randomUUID();
  const now = nowIso();

  // 1. Request blob upload URL
  const reqRes = await fetch(`${STORAGE}/upload/request`, {
    method: "POST",
    headers: { ...authHeaders, "Content-Type": "application/json" },
    body: JSON.stringify([{ ID: docId, Type: "DocumentType", Version: 1 }]),
  });
  await check(reqRes, "Upload request");
  const [info] = await reqRes.json();
  if (!info.Success) throw new Error(`Upload rejected: ${info.Message}`);

  // 2. Build the document ZIP
  const zip = new JSZip();
  zip.file(`${docId}.content`, JSON.stringify({
    extraMetadata: {}, fileType: "pdf", lastOpenedPage: 0,
    lineHeight: -1, margins: 100, pageCount: 1, textScale: 1, transform: {},
  }));
  zip.file(`${docId}.metadata`, JSON.stringify({
    deleted: false, lastModified: now, metadatamodified: false,
    modified: false, parent: "", pinned: false, synced: false,
    type: "DocumentType", version: 1, visibleName: name,
  }));
  zip.file(`${docId}.pagedata`, "");
  zip.file(`${docId}.pdf`, pdfBytes);
  const zipBytes = await zip.generateAsync({ type: "uint8array" });

  // 3. Upload blob
  const blobRes = await fetch(info.BlobURLPut, {
    method: "PUT",
    headers: { "Content-Type": "application/octet-stream" },
    body: Buffer.from(zipBytes),
  });
  await check(blobRes, "Blob upload");

  // 4. Finalise
  const statusRes = await fetch(`${STORAGE}/upload/update-status`, {
    method: "PUT",
    headers: { ...authHeaders, "Content-Type": "application/json" },
    body: JSON.stringify([{
      ID: docId, deleted: false, lastModified: now,
      metadatamodified: true, modified: true, parent: "",
      pinned: false, synced: true, Type: "DocumentType",
      Version: 1, VissibleName: name,   // intentional API typo
    }]),
  });
  await check(statusRes, "Status update");

  return docId;
}

File: sudoku2rm/remarkable.py
"""Unofficial reMarkable cloud API integration.

Authentication flow (one-time):
  1. Visit https://my.remarkable.com/connect/mobile
  2. Copy the 8-character code shown
  3. Run: python run.py setup

Credentials are stored at ~/.config/sudoku2rm/auth.json.
"""

import io
import json
import uuid
import zipfile
from datetime import datetime, timezone
from pathlib import Path

import requests

_AUTH_BASE = "https://my.remarkable.com"
_STORAGE_BASE = (
    "https://document-storage-production-dot-remarkable-production.appspot.com"
    "/document-storage/json/2"
)

CONFIG_DIR = Path.home() / ".config" / "sudoku2rm"
CONFIG_FILE = CONFIG_DIR / "auth.json"


# ── Auth helpers ──────────────────────────────────────────────────────────────

def register_device(code: str) -> None:
    """Register this client with the reMarkable cloud using a one-time code."""
    device_id = str(uuid.uuid4())
    resp = requests.post(
        f"{_AUTH_BASE}/token/json/2/device/new",
        json={
            "code": code.strip().lower(),
            "deviceDesc": "desktop-linux",
            "deviceID": device_id,
        },
        timeout=15,
    )
    _check(resp, "Device registration")

    CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    CONFIG_FILE.write_text(json.dumps({"device_token": resp.text}))
    CONFIG_FILE.chmod(0o600)


def _load_device_token() -> str:
    if not CONFIG_FILE.exists():
        raise RuntimeError(
            "Not authenticated. Run:  python run.py setup"
        )
    return json.loads(CONFIG_FILE.read_text())["device_token"]


def _session_token(device_token: str) -> str:
    resp = requests.post(
        f"{_AUTH_BASE}/token/json/2/user/new",
        headers={"Authorization": f"Bearer {device_token}"},
        timeout=15,
    )
    _check(resp, "Session token refresh")
    return resp.text


def _check(resp: requests.Response, label: str) -> None:
    if not resp.ok:
        raise RuntimeError(f"{label} failed ({resp.status_code}): {resp.text[:200]}")


# ── Document package ──────────────────────────────────────────────────────────

def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def _make_package(doc_id: str, pdf_bytes: bytes, name: str) -> bytes:
    """Wrap a PDF in the reMarkable ZIP document format."""
    content = json.dumps({
        "extraMetadata": {},
        "fileType": "pdf",
        "lastOpenedPage": 0,
        "lineHeight": -1,
        "margins": 100,
        "pageCount": 1,
        "textScale": 1,
        "transform": {},
    })
    metadata = json.dumps({
        "deleted": False,
        "lastModified": _now_iso(),
        "metadatamodified": False,
        "modified": False,
        "parent": "",
        "pinned": False,
        "synced": False,
        "type": "DocumentType",
        "version": 1,
        "visibleName": name,
    })

    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as zf:
        zf.writestr(f"{doc_id}.content", content)
        zf.writestr(f"{doc_id}.metadata", metadata)
        zf.writestr(f"{doc_id}.pagedata", "")
        zf.writestr(f"{doc_id}.pdf", pdf_bytes)
    buf.seek(0)
    return buf.read()


# ── Upload ────────────────────────────────────────────────────────────────────

def upload_pdf(pdf_bytes: bytes, name: str) -> str:
    """Upload a PDF to the reMarkable cloud. Returns the new document ID."""
    device_token = _load_device_token()
    session = _session_token(device_token)
    headers = {"Authorization": f"Bearer {session}"}
    doc_id = str(uuid.uuid4())

    # 1. Request a blob upload URL
    resp = requests.post(
        f"{_STORAGE_BASE}/upload/request",
        headers=headers,
        json=[{"ID": doc_id, "Type": "DocumentType", "Version": 1}],
        timeout=15,
    )
    _check(resp, "Upload request")
    info = resp.json()[0]
    if not info.get("Success"):
        raise RuntimeError(f"Upload request rejected: {info.get('Message')}")

    # 2. PUT the ZIP package to the pre-signed blob URL
    resp = requests.put(
        info["BlobURLPut"],
        data=_make_package(doc_id, pdf_bytes, name),
        headers={"Content-Type": "application/octet-stream"},
        timeout=60,
    )
    _check(resp, "Blob upload")

    # 3. Finalise: update document status
    resp = requests.put(
        f"{_STORAGE_BASE}/upload/update-status",
        headers=headers,
        json=[{
            "ID": doc_id,
            "deleted": False,
            "lastModified": _now_iso(),
            "metadatamodified": True,
            "modified": True,
            "parent": "",
            "pinned": False,
            "synced": True,
            "Type": "DocumentType",
            "Version": 1,
            "VissibleName": name,   # intentional API typo
        }],
        timeout=15,
    )
    _check(resp, "Status update")

    return doc_id
